torch/csrc/jit/passes/graph_fuser.cpp

```diff
@@ -1056,6 +1008,74 @@ struct GraphFuser {
     aliasDb_ = AliasAnalysis(graph_);
   }
 
+  bool canFuseWithConcat(Value* producer, Node* before_check) {
+    if (!isFusable(producer->node())) {
+      return false;
+    }
+    // NB: it is important that this check happens after isFusable, which checks
+    // that the blocks match, and it's not a special node like prim::Param
+    if (!producer->node()->couldMoveBeforeTopologically(
+            before_check, aliasDb_.value())) {
+      return false;
+    }
+    // Fusion groups can be merged with concat's group if and only if
+    // the value they produce isn't already coming from a concat.
+    if (producer->node()->kind() == prim::FusionGroup) {
+      auto subgraph = producer->node()->g(attr::Subgraph);
+      auto* node = subgraph->outputs().at(producer->offset())->node();
+      return node->kind() != prim::FusedConcat;
+    }
+    return true;
+  }
+
+  Node* createFusedConcat(Node* node) {
+    JIT_ASSERT(node->kind() == aten::cat);
+
+    Graph* graph = node->owningGraph();
+    Node* list_construct = node->namedInput(attr::tensors)->node();
+    int64_t dim = node->get<int64_t>(attr::dim).value();
+
+    Node* fused_cat = graph->create(prim::FusedConcat, list_construct->inputs())
+                          ->i_(attr::dim, dim);
+    fused_cat->insertBefore(list_construct);
+    fused_cat->output()->copyMetadata(node->output());
+    node->output()->replaceAllUsesWith(fused_cat->output());
+
+    // NB: this deletes the fused_cat node from the original graph
+    return createSingletonFusionGroup(fused_cat);
+  }
+
+  void fuseConcats() {
+    for (auto it = block_->nodes().rbegin(); it != block_->nodes().rend();
+         ++it) {
+      Node* cat = *it;
+      if (!isFusableCatNode(cat)) {
+        continue;
+      }
+      Node* list_construct = cat->namedInput(attr::tensors)->node();
+      Node* fused_cat = createFusedConcat(cat);
+      it.destroyCurrent();
+      if (list_construct->output()->uses().empty()) {
+        list_construct->destroy();
+      }
+
+      auto sorted_inputs = sortReverseTopological(fused_cat->inputs());
+      size_t input_idx = 0;
+      while (input_idx < sorted_inputs.size()) {
+        Value* input = sorted_inputs[input_idx++];
+        if (!canFuseWithConcat(input, fused_cat)) {
+          continue;
+        }
+        auto maybe_group = tryFuse(fused_cat, input);
```

Comment:
Do we need to update the AliasDb here because we changed the graph? (When should the AliasDb get updated?)

Reply:
Yes. What we do currently is extremely unsafe, but I just keep the current behavior. @suo is working on methods that let us update both AliasDb and the graph. Currently I can't merge my deinplacing PR because AliasDb contains daed pointers after I rewrite it.